File: frontend/src/pages/CRM.js
import React, { useEffect, useMemo, useState } from 'react';
import { useSelector } from 'react-redux';
import api from '../api/axiosConfig';
import toast from 'react-hot-toast';

const LEAD_STATUSES = ['new', 'contacted', 'qualified', 'proposal', 'won', 'lost'];
const TABS = ['overview', 'customers', 'leads', 'tasks'];

const money = (value) => '₹' + Number(value || 0).toLocaleString('en-IN', { maximumFractionDigits: 0 });
const date = (value) => value ? new Date(value).toLocaleDateString('en-IN') : '—';
const fullName = (lead) => [lead.first_name, lead.last_name].filter(Boolean).join(' ');

const Metric = ({ label, value, detail }) => (
  <div className="bg-white rounded-xl shadow-sm border border-gray-100 p-5">
    <p className="text-sm text-gray-500">{label}</p>
    <p className="text-2xl font-bold text-gray-900 mt-2">{value}</p>
    {detail && <p className="text-xs text-gray-500 mt-1">{detail}</p>}
  </div>
);

const CRM = () => {
  const { user } = useSelector((state) => state.auth);
  const [tab, setTab] = useState('overview');
  const [overview, setOverview] = useState(null);
  const [customers, setCustomers] = useState([]);
  const [customerSearch, setCustomerSearch] = useState('');
  const [selectedCustomer, setSelectedCustomer] = useState(null);
  const [leads, setLeads] = useState([]);
  const [activities, setActivities] = useState([]);
  const [loading, setLoading] = useState(true);
  const [leadForm, setLeadForm] = useState({ first_name: '', last_name: '', email: '', phone: '', company: '', source: 'website', value: '', notes: '' });
  const [taskForm, setTaskForm] = useState({ title: '', description: '', due_date: '', customer_id: '', lead_id: '' });

  const loadOverview = async () => {
    const response = await api.getCRMOverview();
    setOverview(response.data);
  };
  const loadCustomers = async () => {
    const response = await api.getCRMCustomers({ search: customerSearch || undefined });
    setCustomers(response.data);
  };
  const loadLeads = async () => {
    const response = await api.getCRMLeads();
    setLeads(response.data);
  };
  const loadActivities = async () => {
    const response = await api.getCRMActivities({ open_only: tab === 'tasks' });
    setActivities(response.data);
  };

  useEffect(() => {
    if (!user?.is_admin) return;
    setLoading(true);
    Promise.all([loadOverview(), loadCustomers(), loadLeads(), loadActivities()])
      .catch((error) => toast.error(error.response?.data?.detail || 'Could not load CRM data'))
      .finally(() => setLoading(false));
  }, [user, customerSearch, tab]);

  if (!user?.is_admin) return <div className="container mx-auto px-4 py-16 text-center text-gray-600">Admin access only</div>;
  if (loading && !overview) return <div className="container mx-auto px-4 py-16 text-center text-gray-600">Loading CRM...</div>;

  const tabClass = (name) => 'px-4 py-2 text-sm font-semibold border-b-2 ' + (tab === name ? 'text-purple-700 border-purple-600' : 'text-gray-500 border-transparent');

  const openCustomer = async (id) => {
    try {
      const response = await api.getCRMCustomer(id);
      setSelectedCustomer(response.data);
    } catch (error) {
      toast.error('Could not load customer');
    }
  };

  const createLead = async (event) => {
    event.preventDefault();
    try {
      await api.createCRMLead({ ...leadForm, value: Number(leadForm.value || 0) });
      setLeadForm({ first_name: '', last_name: '', email: '', phone: '', company: '', source: 'website', value: '', notes: '' });
      await Promise.all([loadLeads(), loadOverview()]);
      toast.success('Lead created');
    } catch (error) {
      toast.error(error.response?.data?.detail || 'Could not create lead');
    }
  };

  const updateLeadStatus = async (lead, status) => {
    try {
      await api.updateCRMLead(lead.id, { status });
      await Promise.all([loadLeads(), loadOverview()]);
    } catch (error) {
      toast.error(error.response?.data?.detail || 'Could not update lead');
    }
  };

  const createTask = async (event) => {
    event.preventDefault();
    if (!taskForm.customer_id && !taskForm.lead_id) {
      toast.error('Select a customer or lead');
      return;
    }
    try {
      await api.createCRMActivity({
        type: 'task',
        title: taskForm.title,
        description: taskForm.description || null,
        due_date: taskForm.due_date ? new Date(taskForm.due_date).toISOString() : null,
        customer_id: taskForm.customer_id ? Number(taskForm.customer_id) : null,
        lead_id: taskForm.lead_id ? Number(taskForm.lead_id) : null,
      });
      setTaskForm({ title: '', description: '', due_date: '', customer_id: '', lead_id: '' });
      await Promise.all([loadActivities(), loadOverview()]);
      toast.success('Task created');
    } catch (error) {
      toast.error(error.response?.data?.detail || 'Could not create task');
    }
  };

  const completeTask = async (activity) => {
    try {
      await api.updateCRMActivity(activity.id, { completed: true });
      await Promise.all([loadActivities(), loadOverview()]);
      if (selectedCustomer) openCustomer(selectedCustomer.customer.id);
    } catch (error) {
      toast.error('Could not complete task');
    }
  };

  const pipelineTotal = useMemo(() => Object.values(overview?.pipeline || {}).reduce((sum, item) => sum + Number(item.value || 0), 0), [overview]);

  return (
    <div className="container mx-auto px-4 py-8 max-w-7xl">
      <div className="flex flex-wrap justify-between items-end gap-4 mb-6">
        <div><p className="text-sm font-semibold text-purple-600 uppercase tracking-wide">Operations</p><h1 className="text-3xl font-bold text-gray-900">Customer Relationship Manager</h1><p className="text-gray-500 mt-1">One workspace for customers, sales follow-up, and order relationships.</p></div>
        <a href="/admin" className="text-sm text-purple-700 hover:underline">Back to store admin →</a>
      </div>
      <div className="flex gap-2 border-b border-gray-200 mb-8 overflow-x-auto">
        {TABS.map((name) => <button key={name} onClick={() => setTab(name)} className={tabClass(name)}>{name[0].toUpperCase() + name.slice(1)}</button>)}
      </div>

      {tab === 'overview' && overview && (
        <div className="space-y-8">
          <div className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-5 gap-4">
            <Metric label="Customers" value={overview.total_customers} detail="Registered shoppers" />
            <Metric label="Orders" value={overview.total_orders} detail="All-time orders" />
            <Metric label="Revenue" value={money(overview.total_revenue)} detail="From existing orders" />
            <Metric label="Active leads" value={overview.active_leads} detail="Not won or lost" />
            <Metric label="Open tasks" value={overview.open_tasks} detail="Needs follow-up" />
          </div>
          <div className="grid grid-cols-1 lg:grid-cols-3 gap-6">
            <div className="lg:col-span-2 bg-white rounded-xl shadow-sm border border-gray-100 p-6">
              <div className="flex justify-between items-center mb-4"><h2 className="font-bold text-lg">Recent orders</h2><button onClick={() => setTab('customers')} className="text-sm text-purple-700">View customers</button></div>
              <div className="divide-y">{overview.recent_orders.map((order) => <div key={order.id} className="py-3 flex flex-wrap justify-between gap-2"><div><p className="font-medium">{order.customer_name}</p><p className="text-xs text-gray-500">{order.order_number} · {date(order.created_at)}</p></div><div className="text-right"><p className="font-semibold">{money(order.total_amount)}</p><span className="text-xs text-gray-500 capitalize">{order.status}</span></div></div>)}{overview.recent_orders.length === 0 && <p className="text-gray-500">No orders yet.</p>}</div>
            </div>
            <div className="bg-white rounded-xl shadow-sm border border-gray-100 p-6"><h2 className="font-bold text-lg mb-4">Pipeline value</h2><p className="text-3xl font-bold text-purple-700 mb-4">{money(pipelineTotal)}</p><div className="space-y-3">{LEAD_STATUSES.map((status) => <div key={status} className="flex justify-between text-sm"><span className="capitalize text-gray-600">{status}</span><span className="font-medium">{money(overview.pipeline?.[status]?.value || 0)} · {overview.pipeline?.[status]?.count || 0}</span></div>)}</div></div>
          </div>
        </div>
      )}

      {tab === 'customers' && (
        <div className="grid grid-cols-1 lg:grid-cols-3 gap-6">
          <div className="lg:col-span-2 bg-white rounded-xl shadow-sm border border-gray-100 p-6"><div className="flex justify-between gap-3 mb-4"><h2 className="font-bold text-lg">Customers</h2><input value={customerSearch} onChange={(event) => setCustomerSearch(event.target.value)} placeholder="Search name or email" className="border rounded-lg px-3 py-2 text-sm w-64" /></div><div className="overflow-x-auto"><table className="w-full text-sm"><thead className="text-left text-gray-500 border-b"><tr><th className="py-3">Customer</th><th className="py-3">Orders</th><th className="py-3">Lifetime value</th><th className="py-3">Last order</th></tr></thead><tbody>{customers.map((customer) => <tr key={customer.id} onClick={() => openCustomer(customer.id)} className="border-b last:border-0 hover:bg-purple-50 cursor-pointer"><td className="py-3"><p className="font-medium">{customer.name}</p><p className="text-xs text-gray-500">{customer.email}</p></td><td className="py-3">{customer.order_count}</td><td className="py-3 font-medium">{money(customer.lifetime_value)}</td><td className="py-3 text-gray-500">{date(customer.last_order_at)}</td></tr>)}</tbody></table>{customers.length === 0 && <p className="text-gray-500 py-6 text-center">No customers found.</p>}</div></div>
          <div className="bg-white rounded-xl shadow-sm border border-gray-100 p-6">{selectedCustomer ? <><div className="flex justify-between"><div><h2 className="font-bold text-lg">{selectedCustomer.customer.name}</h2><p className="text-sm text-gray-500">{selectedCustomer.customer.email}</p><p className="text-sm text-gray-500">{selectedCustomer.customer.phone || 'No phone'} · {selectedCustomer.customer.city || 'No city'}</p></div><button onClick={() => setSelectedCustomer(null)} className="text-gray-400">×</button></div><div className="grid grid-cols-2 gap-3 my-5"><div className="bg-purple-50 rounded-lg p-3"><p className="text-xs text-gray-500">Orders</p><p className="font-bold">{selectedCustomer.customer.order_count}</p></div><div className="bg-purple-50 rounded-lg p-3"><p className="text-xs text-gray-500">Value</p><p className="font-bold">{money(selectedCustomer.customer.lifetime_value)}</p></div></div><h3 className="font-semibold mb-2">Order history</h3><div className="space-y-2 mb-5">{selectedCustomer.orders.map((order) => <div key={order.id} className="flex justify-between text-sm"><span>{order.order_number}<span className="text-gray-400"> · {order.status}</span></span><span>{money(order.total_amount)}</span></div>)}{selectedCustomer.orders.length === 0 && <p className="text-sm text-gray-500">No orders yet.</p>}</div><h3 className="font-semibold mb-2">Activities</h3><div className="space-y-2">{selectedCustomer.activities.map((activity) => <div key={activity.id} className="text-sm"><p className="font-medium">{activity.title}</p><p className="text-xs text-gray-500">{activity.type} · {activity.completed ? 'Completed' : 'Open'}</p></div>)}{selectedCustomer.activities.length === 0 && <p className="text-sm text-gray-500">No activities yet.</p>}</div></> : <p className="text-gray-500 text-sm">Select a customer to see order history and CRM activity.</p>}</div>
        </div>
      )}

      {tab === 'leads' && (
        <div className="grid grid-cols-1 lg:grid-cols-3 gap-6"><form onSubmit={createLead} className="bg-white rounded-xl shadow-sm border border-gray-100 p-6 space-y-3"><h2 className="font-bold text-lg mb-4">Add lead</h2><div className="grid grid-cols-2 gap-3"><input required value={leadForm.first_name} onChange={(e) => setLeadForm({ ...leadForm, first_name: e.target.value })} placeholder="First name" className="border rounded-lg px-3 py-2" /><input value={leadForm.last_name} onChange={(e) => setLeadForm({ ...leadForm, last_name: e.target.value })} placeholder="Last name" className="border rounded-lg px-3 py-2" /></div><input type="email" value={leadForm.email} onChange={(e) => setLeadForm({ ...leadForm, email: e.target.value })} placeholder="Email" className="border rounded-lg px-3 py-2 w-full" /><input value={leadForm.phone} onChange={(e) => setLeadForm({ ...leadForm, phone: e.target.value })} placeholder="Phone" className="border rounded-lg px-3 py-2 w-full" /><input value={leadForm.company} onChange={(e) => setLeadForm({ ...leadForm, company: e.target.value })} placeholder="Company or group" className="border rounded-lg px-3 py-2 w-full" /><div className="grid grid-cols-2 gap-3"><select value={leadForm.source} onChange={(e) => setLeadForm({ ...leadForm, source: e.target.value })} className="border rounded-lg px-3 py-2"><option value="website">Website</option><option value="referral">Referral</option><option value="social">Social</option><option value="offline">Offline</option></select><input type="number" min="0" value={leadForm.value} onChange={(e) => setLeadForm({ ...leadForm, value: e.target.value })} placeholder="Value ₹" className="border rounded-lg px-3 py-2" /></div><textarea value={leadForm.notes} onChange={(e) => setLeadForm({ ...leadForm, notes: e.target.value })} placeholder="Notes" rows="3" className="border rounded-lg px-3 py-2 w-full" /><button className="btn-primary w-full">Create lead</button></form><div className="lg:col-span-2 bg-white rounded-xl shadow-sm border border-gray-100 p-6"><h2 className="font-bold text-lg mb-4">Sales pipeline</h2><div className="space-y-3">{leads.map((lead) => <div key={lead.id} className="border rounded-lg p-4 flex flex-wrap justify-between gap-3"><div><p className="font-semibold">{fullName(lead)}</p><p className="text-sm text-gray-500">{lead.email || 'No email'}{lead.company ? ' · ' + lead.company : ''}</p><p className="text-xs text-gray-400 mt-1">{lead.source} · {money(lead.value)}</p></div><select value={lead.status} onChange={(e) => updateLeadStatus(lead, e.target.value)} className="border rounded-lg px-3 py-2 text-sm">{LEAD_STATUSES.map((status) => <option key={status} value={status}>{status}</option>)}</select></div>)}{leads.length === 0 && <p className="text-gray-500">No leads yet. Add your first prospect.</p>}</div></div></div>
      )}

      {tab === 'tasks' && (
        <div className="grid grid-cols-1 lg:grid-cols-3 gap-6"><form onSubmit={createTask} className="bg-white rounded-xl shadow-sm border border-gray-100 p-6 space-y-3"><h2 className="font-bold text-lg mb-4">Create follow-up task</h2><input required value={taskForm.title} onChange={(e) => setTaskForm({ ...taskForm, title: e.target.value })} placeholder="Task title" className="border rounded-lg px-3 py-2 w-full" /><textarea value={taskForm.description} onChange={(e) => setTaskForm({ ...taskForm, description: e.target.value })} placeholder="Details" rows="3" className="border rounded-lg px-3 py-2 w-full" /><input type="date" value={taskForm.due_date} onChange={(e) => setTaskForm({ ...taskForm, due_date: e.target.value })} className="border rounded-lg px-3 py-2 w-full" /><select value={taskForm.customer_id} onChange={(e) => setTaskForm({ ...taskForm, customer_id: e.target.value, lead_id: '' })} className="border rounded-lg px-3 py-2 w-full"><option value="">Select customer</option>{customers.map((customer) => <option key={customer.id} value={customer.id}>{customer.name}</option>)}</select><select value={taskForm.lead_id} onChange={(e) => setTaskForm({ ...taskForm, lead_id: e.target.value, customer_id: '' })} className="border rounded-lg px-3 py-2 w-full"><option value="">Or select lead</option>{leads.map((lead) => <option key={lead.id} value={lead.id}>{fullName(lead)}</option>)}</select><button className="btn-primary w-full">Create task</button></form><div className="lg:col-span-2 bg-white rounded-xl shadow-sm border border-gray-100 p-6"><h2 className="font-bold text-lg mb-4">Open follow-ups</h2><div className="space-y-3">{activities.map((activity) => <div key={activity.id} className="border rounded-lg p-4 flex justify-between gap-4"><div><p className="font-semibold">{activity.title}</p><p className="text-sm text-gray-500">{activity.customer_name || activity.lead_name || 'Unassigned'} · due {date(activity.due_date)}</p><p className="text-sm text-gray-600 mt-1">{activity.description || 'No details'}</p></div><button onClick={() => completeTask(activity)} className="text-sm text-green-700 whitespace-nowrap">Complete</button></div>)}{activities.length === 0 && <p className="text-gray-500">No open tasks.</p>}</div></div></div>
      )}
    </div>
  );
};

export default CRM;

File: backend/app/routes/crm.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, or_
from sqlalchemy.orm import Session

from app import models, schemas
from app.database import get_db
from app.utils.auth import get_current_admin_user

router = APIRouter()

LEAD_STATUSES = {"new", "contacted", "qualified", "proposal", "won", "lost"}
ACTIVITY_TYPES = {"note", "call", "email", "meeting", "task"}


def _name(user):
    return " ".join(part for part in [user.first_name, user.last_name] if part).strip() or user.username


def _customer_summary(user, stats):
    value = stats.get(user.id, {})
    return {
        "id": user.id,
        "name": _name(user),
        "email": user.email,
        "phone": user.phone,
        "city": user.city,
        "created_at": user.created_at,
        "order_count": value.get("order_count", 0),
        "lifetime_value": float(value.get("lifetime_value", 0) or 0),
        "last_order_at": value.get("last_order_at"),
    }


def _activity_response(activity, db):
    customer = db.query(models.User).filter(models.User.id == activity.customer_id).first() if activity.customer_id else None
    lead = db.query(models.CRMLead).filter(models.CRMLead.id == activity.lead_id).first() if activity.lead_id else None
    return {
        "id": activity.id,
        "customer_id": activity.customer_id,
        "lead_id": activity.lead_id,
        "created_by_id": activity.created_by_id,
        "type": activity.type,
        "title": activity.title,
        "description": activity.description,
        "due_date": activity.due_date,
        "completed": activity.completed,
        "created_at": activity.created_at,
        "updated_at": activity.updated_at,
        "customer_name": _name(customer) if customer else None,
        "lead_name": " ".join(part for part in [lead.first_name, lead.last_name] if part).strip() if lead else None,
    }


@router.get("/crm/overview", response_model=schemas.CRMOverviewResponse)
def get_overview(db: Session = Depends(get_db), current_user: models.User = Depends(get_current_admin_user)):
    total_customers = db.query(models.User).filter(models.User.is_admin == False).count()
    total_orders = db.query(models.Order).count()
    total_revenue = db.query(func.coalesce(func.sum(models.Order.total_amount), 0)).scalar() or 0
    active_leads = db.query(models.CRMLead).filter(models.CRMLead.status.notin_(["won", "lost"])).count()
    open_tasks = db.query(models.CRMActivity).filter(models.CRMActivity.type == "task", models.CRMActivity.completed == False).count()

    pipeline_rows = db.query(models.CRMLead.status, func.count(models.CRMLead.id), func.coalesce(func.sum(models.CRMLead.value), 0)).group_by(models.CRMLead.status).all()
    pipeline = {status: {"count": count, "value": float(value or 0)} for status, count, value in pipeline_rows}

    recent_rows = db.query(models.Order, models.User).join(models.User, models.User.id == models.Order.user_id).order_by(models.Order.created_at.desc()).limit(8).all()
    recent_orders = [{
        "id": order.id,
        "order_number": order.order_number,
        "customer_name": _name(user),
        "total_amount": order.total_amount,
        "status": order.status,
        "created_at": order.created_at,
    } for order, user in recent_rows]

    return {
        "total_customers": total_customers,
        "total_orders": total_orders,
        "total_revenue": float(total_revenue),
        "active_leads": active_leads,
        "open_tasks": open_tasks,
        "recent_orders": recent_orders,
        "pipeline": pipeline,
    }


@router.get("/crm/customers", response_model=list[schemas.CRMCustomerSummary])
def list_customers(
    search: Optional[str] = Query(None, max_length=100),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_admin_user),
):
    query = db.query(models.User).filter(models.User.is_admin == False)
    if search:
        term = "%" + search.strip() + "%"
        query = query.filter(or_(models.User.username.ilike(term), models.User.email.ilike(term), models.User.first_name.ilike(term), models.User.last_name.ilike(term)))
    users = query.order_by(models.User.created_at.desc()).offset(skip).limit(limit).all()
    user_ids = [user.id for user in users]
    stats = {}
    if user_ids:
        rows = db.query(models.Order.user_id, func.count(models.Order.id), func.coalesce(func.sum(models.Order.total_amount), 0), func.max(models.Order.created_at)).filter(models.Order.user_id.in_(user_ids)).group_by(models.Order.user_id).all()
        stats = {user_id: {"order_count": count, "lifetime_value": value, "last_order_at": last_order_at} for user_id, count, value, last_order_at in rows}
    return [_customer_summary(user, stats) for user in users]


@router.get("/crm/customers/{customer_id}", response_model=schemas.CRMCustomerDetail)
def get_customer(customer_id: int, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_admin_user)):
    customer = db.query(models.User).filter(models.User.id == customer_id, models.User.is_admin == False).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    rows = db.query(models.Order.user_id, func.count(models.Order.id), func.coalesce(func.sum(models.Order.total_amount), 0), func.max(models.Order.created_at)).filter(models.Order.user_id == customer_id).group_by(models.Order.user_id).all()
    stats = {customer_id: {"order_count": rows[0][1], "lifetime_value": rows[0][2], "last_order_at": rows[0][3]}} if rows else {}
    orders = db.query(models.Order).filter(models.Order.user_id == customer_id).order_by(models.Order.created_at.desc()).limit(50).all()
    activities = db.query(models.CRMActivity).filter(models.CRMActivity.customer_id == customer_id).order_by(models.CRMActivity.created_at.desc()).limit(50).all()
    return {
        "customer": _customer_summary(customer, stats),
        "orders": [{"id": order.id, "order_number": order.order_number, "total_amount": order.total_amount, "status": order.status, "payment_status": order.payment_status, "created_at": order.created_at} for order in orders],
        "activities": [_activity_response(activity, db) for activity in activities],
    }


@router.get("/crm/leads", response_model=list[schemas.CRMLeadResponse])
def list_leads(
    status: Optional[str] = Query(None),
    search: Optional[str] = Query(None, max_length=100),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_admin_user),
):
    query = db.query(models.CRMLead)
    if status:
        query = query.filter(models.CRMLead.status == status)
    if search:
        term = "%" + search.strip() + "%"
        query = query.filter(or_(models.CRMLead.first_name.ilike(term), models.CRMLead.last_name.ilike(term), models.CRMLead.email.ilike(term), models.CRMLead.company.ilike(term)))
    return query.order_by(models.CRMLead.created_at.desc()).all()


@router.post("/crm/leads", response_model=schemas.CRMLeadResponse, status_code=201)
def create_lead(lead: schemas.CRMLeadCreate, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_admin_user)):
    if lead.status not in LEAD_STATUSES:
        raise HTTPException(status_code=400, detail="Invalid lead status")
    db_lead = models.CRMLead(**lead.model_dump())
    db.add(db_lead)
    db.commit()
    db.refresh(db_lead)
    return db_lead


@router.patch("/crm/leads/{lead_id}", response_model=schemas.CRMLeadResponse)
def update_lead(lead_id: int, lead: schemas.CRMLeadUpdate, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_admin_user)):
    db_lead = db.query(models.CRMLead).filter(models.CRMLead.id == lead_id).first()
    if not db_lead:
        raise HTTPException(status_code=404, detail="Lead not found")
    updates = lead.model_dump(exclude_unset=True)
    if "status" in updates and updates["status"] not in LEAD_STATUSES:
        raise HTTPException(status_code=400, detail="Invalid lead status")
    for key, value in updates.items():
        setattr(db_lead, key, value)
    db.commit()
    db.refresh(db_lead)
    return db_lead


@router.get("/crm/activities", response_model=list[schemas.CRMActivityResponse])
def list_activities(
    customer_id: Optional[int] = Query(None),
    lead_id: Optional[int] = Query(None),
    open_only: bool = Query(False),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_admin_user),
):
    query = db.query(models.CRMActivity)
    if customer_id:
        query = query.filter(models.CRMActivity.customer_id == customer_id)
    if lead_id:
        query = query.filter(models.CRMActivity.lead_id == lead_id)
    if open_only:
        query = query.filter(models.CRMActivity.completed == False)
    activities = query.order_by(models.CRMActivity.completed.asc(), models.CRMActivity.due_date.asc(), models.CRMActivity.created_at.desc()).limit(100).all()
    return [_activity_response(activity, db) for activity in activities]


@router.post("/crm/activities", response_model=schemas.CRMActivityResponse, status_code=201)
def create_activity(activity: schemas.CRMActivityCreate, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_admin_user)):
    if activity.type not in ACTIVITY_TYPES:
        raise HTTPException(status_code=400, detail="Invalid activity type")
    if not activity.customer_id and not activity.lead_id:
        raise HTTPException(status_code=400, detail="An activity needs a customer or lead")
    if activity.customer_id and not db.query(models.User).filter(models.User.id == activity.customer_id, models.User.is_admin == False).first():
        raise HTTPException(status_code=404, detail="Customer not found")
    if activity.lead_id and not db.query(models.CRMLead).filter(models.CRMLead.id == activity.lead_id).first():
        raise HTTPException(status_code=404, detail="Lead not found")
    db_activity = models.CRMActivity(**activity.model_dump(), created_by_id=current_user.id)
    db.add(db_activity)
    db.commit()
    db.refresh(db_activity)
    return _activity_response(db_activity, db)


@router.patch("/crm/activities/{activity_id}", response_model=schemas.CRMActivityResponse)
def update_activity(activity_id: int, activity: schemas.CRMActivityUpdate, db: Session = Depends(get_db), current_user: models.User = Depends(get_current_admin_user)):
    db_activity = db.query(models.CRMActivity).filter(models.CRMActivity.id == activity_id).first()
    if not db_activity:
        raise HTTPException(status_code=404, detail="Activity not found")
    updates = activity.model_dump(exclude_unset=True)
    if "type" in updates and updates["type"] not in ACTIVITY_TYPES:
        raise HTTPException(status_code=400, detail="Invalid activity type")
    for key, value in updates.items():
        setattr(db_activity, key, value)
    db.commit()
    db.refresh(db_activity)
    return _activity_response(db_activity, db)
